File: src/audit.js
function lineOf(text, index) {
  return text.slice(0, index).split(/\r?\n/).length;
}

export function auditHtml(html) {
  const issues = [];

  const htmlTagMatch = html.match(/<html\b[^>]*>/i);
  if (htmlTagMatch) {
    const hasLang = /\blang\s*=\s*["'][^"']+["']/i.test(htmlTagMatch[0]);
    if (!hasLang) {
      issues.push({
        rule: "page-lang",
        severity: "medium",
        message: "Tag <html> senza attributo lang.",
        line: lineOf(html, htmlTagMatch.index ?? 0)
      });
    }
  }

  const h1Regex = /<h1\b[^>]*>[\s\S]*?<\/h1>/gi;
  const h1Count = [...html.matchAll(h1Regex)].length;
  if (h1Count === 0) {
    issues.push({
      rule: "heading-h1",
      severity: "medium",
      message: "Pagina senza heading principale <h1>.",
      line: 1
    });
  }

  const imgRegex = /<img\b[^>]*>/gi;
  for (const match of html.matchAll(imgRegex)) {
    const tag = match[0];
    const hasAlt = /\balt\s*=\s*["'][^"']*["']/i.test(tag);
    if (!hasAlt) {
      issues.push({
        rule: "img-alt",
        severity: "high",
        message: "Immagine senza attributo alt.",
        line: lineOf(html, match.index ?? 0)
      });
    }
  }

  const inputRegex = /<input\b[^>]*>/gi;
  for (const match of html.matchAll(inputRegex)) {
    const tag = match[0];
    const idMatch = tag.match(/\bid\s*=\s*["']([^"']+)["']/i);
    const hasAriaLabel = /\baria-label\s*=\s*["'][^"']+["']/i.test(tag);
    let hasLabel = false;

    if (idMatch?.[1]) {
      const id = idMatch[1].replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
      const labelRegex = new RegExp(`<label\\b[^>]*for\\s*=\\s*["']${id}["'][^>]*>`, "i");
      hasLabel = labelRegex.test(html);
    }

    if (!hasLabel && !hasAriaLabel) {
      issues.push({
        rule: "input-label",
        severity: "high",
        message: "Input senza label associata o aria-label.",
        line: lineOf(html, match.index ?? 0)
      });
    }
  }

  const buttonRegex = /<button\b[^>]*>([\s\S]*?)<\/button>/gi;
  for (const match of html.matchAll(buttonRegex)) {
    const tag = match[0];
    const innerText = (match[1] || "").replace(/<[^>]*>/g, "").trim();
    const hasAriaLabel = /\baria-label\s*=\s*["'][^"']+["']/i.test(tag);
    if (!innerText && !hasAriaLabel) {
      issues.push({
        rule: "button-name",
        severity: "medium",
        message: "Button senza testo leggibile o aria-label.",
        line: lineOf(html, match.index ?? 0)
      });
    }
  }

  const anchorRegex = /<a\b[^>]*>([\s\S]*?)<\/a>/gi;
  for (const match of html.matchAll(anchorRegex)) {
    const tag = match[0];
    const innerText = (match[1] || "").replace(/<[^>]*>/g, "").trim();
    const hasAriaLabel = /\baria-label\s*=\s*["'][^"']+["']/i.test(tag);
    if (!innerText && !hasAriaLabel) {
      issues.push({
        rule: "link-name",
        severity: "medium",
        message: "Link senza testo leggibile o aria-label.",
        line: lineOf(html, match.index ?? 0)
      });
    }
  }

  return {
    score: Math.max(0, 100 - issues.length * 15),
    issueCount: issues.length,
    issues
  };
}

File: python-app/conforme_toolkit/audit.py
import re


def _line_of(text: str, index: int) -> int:
    return text[:index].count("\n") + 1


def audit_html(html: str) -> dict:
    issues = []

    html_tag_match = re.search(r"<html\b[^>]*>", html, flags=re.IGNORECASE)
    if html_tag_match:
        has_lang = re.search(r'\blang\s*=\s*["\'][^"\']+["\']', html_tag_match.group(0), flags=re.IGNORECASE)
        if not has_lang:
            issues.append(
                {
                    "rule": "page-lang",
                    "severity": "medium",
                    "message": "Tag <html> senza attributo lang.",
                    "line": _line_of(html, html_tag_match.start()),
                }
            )

    h1_count = len(re.findall(r"<h1\b[^>]*>[\s\S]*?</h1>", html, flags=re.IGNORECASE))
    if h1_count == 0:
        issues.append(
            {
                "rule": "heading-h1",
                "severity": "medium",
                "message": "Pagina senza heading principale <h1>.",
                "line": 1,
            }
        )

    for match in re.finditer(r"<img\b[^>]*>", html, flags=re.IGNORECASE):
        tag = match.group(0)
        has_alt = re.search(r'\balt\s*=\s*["\'][^"\']*["\']', tag, flags=re.IGNORECASE)
        if not has_alt:
            issues.append(
                {
                    "rule": "img-alt",
                    "severity": "high",
                    "message": "Immagine senza attributo alt.",
                    "line": _line_of(html, match.start()),
                }
            )

    for match in re.finditer(r"<input\b[^>]*>", html, flags=re.IGNORECASE):
        tag = match.group(0)
        id_match = re.search(r'\bid\s*=\s*["\']([^"\']+)["\']', tag, flags=re.IGNORECASE)
        has_aria_label = re.search(r'\baria-label\s*=\s*["\'][^"\']+["\']', tag, flags=re.IGNORECASE)
        has_label = False

        if id_match:
            field_id = re.escape(id_match.group(1))
            label_regex = rf"<label\b[^>]*for\s*=\s*['\"]{field_id}['\"][^>]*>"
            has_label = re.search(label_regex, html, flags=re.IGNORECASE) is not None

        if not has_label and not has_aria_label:
            issues.append(
                {
                    "rule": "input-label",
                    "severity": "high",
                    "message": "Input senza label associata o aria-label.",
                    "line": _line_of(html, match.start()),
                }
            )

    for match in re.finditer(r"<button\b[^>]*>([\s\S]*?)</button>", html, flags=re.IGNORECASE):
        tag = match.group(0)
        inner_text = re.sub(r"<[^>]*>", "", match.group(1)).strip()
        has_aria_label = re.search(r'\baria-label\s*=\s*["\'][^"\']+["\']', tag, flags=re.IGNORECASE)
        if not inner_text and not has_aria_label:
            issues.append(
                {
                    "rule": "button-name",
                    "severity": "medium",
                    "message": "Button senza testo leggibile o aria-label.",
                    "line": _line_of(html, match.start()),
                }
            )

    for match in re.finditer(r"<a\b[^>]*>([\s\S]*?)</a>", html, flags=re.IGNORECASE):
        tag = match.group(0)
        inner_text = re.sub(r"<[^>]*>", "", match.group(1)).strip()
        has_aria_label = re.search(r'\baria-label\s*=\s*["\'][^"\']+["\']', tag, flags=re.IGNORECASE)
        if not inner_text and not has_aria_label:
            issues.append(
                {
                    "rule": "link-name",
                    "severity": "medium",
                    "message": "Link senza testo leggibile o aria-label.",
                    "line": _line_of(html, match.start()),
                }
            )

    return {
        "score": max(0, 100 - len(issues) * 15),
        "issueCount": len(issues),
        "issues": issues,
    }
